### PR TITLE
Change handle fragment - DataSources + UseCases + Skeleton ViewModel/Fragment (Part 1) 

### PR DESCRIPTION
**Pull request by BradleyWire**
## What's new in this PR?

This is the fragment that opens when the user wants to update their handle through the account settings. 
#### APK
[Download build #922](http://10.10.124.11:8080/job/Pull%20Request%20Builder/922/artifact/build/artifact/wire-dev-PR2549-922.apk)
[Download build #935](http://10.10.124.11:8080/job/Pull%20Request%20Builder/935/artifact/build/artifact/wire-dev-PR2549-935.apk)
[Download build #943](http://10.10.124.11:8080/job/Pull%20Request%20Builder/943/artifact/build/artifact/wire-dev-PR2549-943.apk)
[Download build #948](http://10.10.124.11:8080/job/Pull%20Request%20Builder/948/artifact/build/artifact/wire-dev-PR2549-948.apk)
[Download build #963](http://10.10.124.11:8080/job/Pull%20Request%20Builder/963/artifact/build/artifact/wire-dev-PR2549-963.apk)
[Download build #964](http://10.10.124.11:8080/job/Pull%20Request%20Builder/964/artifact/build/artifact/wire-dev-PR2549-964.apk)

### CODE REVIEW

File: app/src/main/res-v2/layout/fragment_edit_handle_dialog.xml [hunk 1]
```diff
@@ -0,0 +1,151 @@
+<?xml version="1.0" encoding="utf-8"?><!--
```
Comment (BradleyWire) [inline app/src/main/res-v2/layout/fragment_edit_handle_dialog.xml]: This file will be changed.

File: app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import com.waz.zclient.core.functional.getOrElse
+import com.waz.zclient.core.usecase.UseCase
+import com.waz.zclient.user.data.UsersRepository
+
+object HandleTooLongError : ValidateHandleState()
+object HandleTooShortError : ValidateHandleState()
+object HandleInvalidError : ValidateHandleState()
+object HandleExistsAlreadyError : ValidateHandleState()
+
+sealed class ValidateHandleState : FeatureFailure()
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt]: Will rename this to ValidateHandleError

File: app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountFragment.kt [hunk 3]
```diff
@@ -132,7 +136,17 @@ class SettingsAccountFragment : Fragment() {
         Toast.makeText(requireContext(), errorMessage, Toast.LENGTH_LONG).show()
     }
 
-    private fun showEditDialog(title: String, defaultValue: String, updateFunc: (String) -> Unit) {
+    private fun showEditHandleDialog(updateHandle: (String) -> Unit) {
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountFragment.kt]: This is never used

File: app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountFragment.kt [hunk 4]
```diff
@@ -69,7 +73,7 @@ class SettingsAccountFragment : Fragment() {
         preferences_account_handle.setOnClickListener {
             val title = getString(R.string.pref__account_action__dialog__change_username__title)
             val defaultValue = preferences_account_handle_title.text.toString()
-            showEditDialog(title, defaultValue) { settingsAccountViewModel.updateHandle(it) }
+            showGenericEditDialog(title, defaultValue) { settingsAccountViewModel.updateHandle(it) }
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountFragment.kt]: You should call `showEditHandleDialog` here

File: app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountFragment.kt (hunk 3 above)
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/settings/account/SettingsAccountFragment.kt]: This PR is a work in progress at the minute.

File: app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersNetworkService.kt [hunk 5]
```diff
@@ -31,6 +24,19 @@ interface UsersNetworkService {
 
     @PUT("$SELF$PHONE")
     suspend fun changePhone(@Body phone: ChangePhoneRequest): Response<Unit>
+
+    @GET("$USERS$HANDLES/{newHandle}")
+    suspend fun doesHandleExist(@Path("handle") newHandle: String): Response<Unit>
+
+    companion object {
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersNetworkService.kt]: 👍

File: app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/GetHandleUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+
+@InternalCoroutinesApi
+@ExperimentalCoroutinesApi
+class GetHandleUseCase(private val userRepository: UsersRepository) : ObservableUseCase<String, Unit>() {
+
+    override suspend fun run(params: Unit): Flow<String> = flow {
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/GetHandleUseCase.kt]: 👍

File: app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package com.waz.zclient.user.domain.usecase.handle
+
+import com.waz.zclient.core.exception.Failure
+import com.waz.zclient.core.functional.Either
+import com.waz.zclient.core.usecase.UseCase
+
+class ValidateHandleUseCase : UseCase<String, ValidateHandleParams>() {
+
+    override suspend fun run(params: ValidateHandleParams): Either<Failure, String> =
+        if (!params.newHandle.matches(HANDLE_REGEX)) {
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt]: You can extract that to a private method `validateHandle(params)`

File: app/src/main/scala/com/waz/zclient/WireApplication.scala [hunk 8]
```diff
@@ -129,6 +130,7 @@ object WireApplication extends DerivedLogTag {
     bind [AudioManager]         to ctx.getSystemService(Context.AUDIO_SERVICE).asInstanceOf[AudioManager]
     bind [SensorManager]        to ctx.getSystemService(Context.SENSOR_SERVICE).asInstanceOf[SensorManager]
     bind [NotificationManager]  to ctx.getSystemService(Context.NOTIFICATION_SERVICE).asInstanceOf[NotificationManager]
+    bind [ShortcutManager]      to ctx.getSystemService(Context.SHORTCUT_SERVICE).asInstanceOf[ShortcutManager]
```
Comment (mejdoo) [inline app/src/main/scala/com/waz/zclient/WireApplication.scala]: out of scope I guess no?

File: app/src/main/scala/com/waz/zclient/pages/main/MainPhoneFragment.scala [hunk 9]
```diff
@@ -214,7 +214,7 @@ class MainPhoneFragment extends FragmentHelper
     //Only for internal builds for now, will be for production when approved by QA.
     if (Build.VERSION.SDK_INT > 25 && BuildConfig.FLAVOR.equals("internal")) {
       val shortcuts = new Shortcuts()
-      val shortcutManager = getActivity.getSystemService(classOf[ShortcutManager])
+      val shortcutManager = inject[ShortcutManager]
```
Comment (mejdoo) [inline app/src/main/scala/com/waz/zclient/pages/main/MainPhoneFragment.scala]: same here

Comment (BradleyWire) [reply app/src/main/scala/com/waz/zclient/pages/main/MainPhoneFragment.scala]: I didn't want to raise a tiny PR for this. But, happy to do so.

File: app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/EditHandleFragment.kt [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        override fun afterTextChanged(s: Editable?) {
+            editHandleFragmentViewModel.afterHandleTextChanged(s.toString())
+        }
+
+        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
+            editHandleFragmentViewModel.beforeHandleTextChanged(s.toString())
+        }
+
+        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+            // Do nothing
```
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/EditHandleFragment.kt]: Good catch, i'll sort it.

File: app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt (hunk 7 above)
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt]: done

File: app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersRemoteDataSource.kt [hunk 11]
```diff
@@ -16,4 +17,18 @@ class UsersRemoteDataSource constructor(private val usersNetworkService: UsersNe
     suspend fun changeEmail(email: String): Either<Failure, Any> = requestApi { usersNetworkService.changeEmail(ChangeEmailRequest(email)) }
 
     suspend fun changePhone(phone: String): Either<Failure, Any> = requestApi { usersNetworkService.changePhone(ChangePhoneRequest(phone)) }
+
+    suspend fun doesHandleExist(newHandle: String): Either<Failure, ValidateHandleSuccess> =
+        when (usersNetworkService.doesHandleExist(newHandle).code()) {
```
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersRemoteDataSource.kt]: We're simply abiding by the rules of the network API. If we integrate the database in some way, it won't be consistent behaviour. I discussed this with @android10 and he agreed.

File: app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/EditHandleFragment.kt [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
+        inflater.inflate(R.layout.fragment_edit_handle_dialog, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        initListeners()
+    }
+
+    private fun initListeners() {
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/EditHandleFragment.kt]: I approve this naming when it's a smaller class with limited functionality, but, it's better to break out further into more private methods when it's a full fragment or activity as this method can become quite big.

File: app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersRemoteDataSource.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (mejdoo) [reply app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersRemoteDataSource.kt]: ok makes sense

File: app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+            Either.Left(HandleInvalidError)
+        } else {
+            when {
+                params.newHandle.length > HANDLE_MAX_LENGTH -> Either.Left(HandleTooLongError)
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/user/domain/usecase/handle/ValidateHandleUseCase.kt]: You can also extract two private methods `handleMinimumLengthValid` and `handleMaximumLengthValid`and unit test them same as `handleCharactersValid` 👍

File: app/src/test/kotlin/com/waz/zclient/user/data/source/remote/UserRemoteDataSourceTest.kt [hunk 14]
```diff
@@ -231,4 +223,64 @@ class UserRemoteDataSourceTest : UnitTest() {
...
+        usersRemoteDataSource.doesHandleExist(TEST_HANDLE)
+
+        verify(usersNetworkService).doesHandleExist(eq(TEST_HANDLE))
+
+        if (cancelled) {
+            cancel(CancellationException(TEST_EXCEPTION_MESSAGE))
+            delay(CANCELLATION_DELAY)
+        }
+
+        usersRemoteDataSource.doesHandleExist(TEST_HANDLE).isLeft shouldBe isError
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/user/data/source/remote/UserRemoteDataSourceTest.kt]: maybe we can check error type too? (HandleTooLong, etc.)

File: app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/EditHandleFragment.kt [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    interface HandleChangedListener {
+        fun onHandleChanged(handle: String)
+    }
+
+    companion object {
+        private const val CURRENT_HANDLE_BUNDLE_KEY = "currentHandleBundleKey"
+
+        fun newInstance(currentHandle: String?, handleChangedListener: HandleChangedListener):
```
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/EditHandleFragment.kt]: I've just realised we don't even need the listener as we're updating the data within the viewmodel in part 2, so i've removed it here.

File: app/src/main/res-v2/layout/fragment_edit_handle_dialog.xml [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+                        android:drawableStart="@drawable/empty_at"
+                        android:inputType="textNoSuggestions|textVisiblePassword"
+                        android:maxLines="1"
+                        android:paddingBottom="@dimen/prefs__username_change__edit_text_bottom_padding"
+                        android:visibility="visible" />
+
+                </com.google.android.material.textfield.TextInputLayout>
+
+                <androidx.appcompat.widget.AppCompatTextView
+                    android:id="@+id/act__at"
```
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/res-v2/layout/fragment_edit_handle_dialog.xml]: XML will be looked at as a whole in the end when we properly define our conventions. So we've moved it to a different resources sourceSet and added new files so we can still remove the old ones and keep the new ones.

File: app/src/test/kotlin/com/waz/zclient/user/data/source/remote/UserRemoteDataSourceTest.kt (hunk 14 above)
Comment (BradleyWire) [reply app/src/test/kotlin/com/waz/zclient/user/data/source/remote/UserRemoteDataSourceTest.kt]: yes, good shout.

File: app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersNetworkService.kt [hunk 17]
```diff
@@ -31,6 +24,19 @@ interface UsersNetworkService {
 
     @PUT("$SELF$PHONE")
     suspend fun changePhone(@Body phone: ChangePhoneRequest): Response<Unit>
+
+    @GET("$USERS$HANDLES/{newHandle}")
+    suspend fun doesHandleExist(@Path("handle") newHandle: String): Response<Unit>
```
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersNetworkService.kt]: newHandle is the value being passed and handle is the endpoint the path is pointed to

File: app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/HandleTextChangedListener.kt [hunk 18]
```diff
@@ -0,0 +1,8 @@
+package com.waz.zclient.settings.account.edithandle
+
+import android.text.TextWatcher
+
+//Temporary naming. Needs to reconsider the name.
+interface HandleTextChangedListener : TextWatcher {
```
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/settings/account/edithandle/HandleTextChangedListener.kt]: Will do

File: app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersNetworkService.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/user/data/source/remote/UsersNetworkService.kt]: Ah my bad, you're right. Changed it.

**Pull request merged**